src/Query.jl

```diff
@@ -0,0 +1,243 @@
+module Query
+
+export From, Where, Select, AbstractQueryFormatter, SimpleQueryFormatter, NamedQueryFormatter, DFQueryFormatter
+
+using ..ACSetInterface, ..Schemas
+using MLStyle
+using DataFrames: DataFrame
+using StructEquality
+
+to_name(x) = x
+to_name(x::Pair) = Symbol("$(x.second)$(x.first)")
+to_name(x::Val{T}) where T = Symbol("Val_$T")
+
+""" This function lifts a singleton into an iterator. This prevents branches in the control flow.
+```
+a=1; @test [k for k ∈ iterable(a)] == [1]
+a=[1,2,3]; @test [k for k ∈ iterable(a)] == [1,2,3]
+```
+"""
+function iterable(x::T) where T
+  S = T <: AbstractVector ? eltype(T) : T
+  [S[]; x]
+end
+
+function Base.get(acset::ACSet, select::Symbol, idx=Colon(); schema=acset_schema(acset))
+  val = select ∈ objects(schema) ? parts(acset, select) : subpart(acset, select)
+  val[idx]
+end
+
+function Base.get(acset, selects::Vector{Symbol}, idx=Colon(); kwargs...)
+  zip(get.(Ref(acset), selects, idx; kwargs...)...)
+end
+
+abstract type AbstractCondition end
+
+@struct_hash_equal struct WhereCondition <: AbstractCondition
+  lhs
+  op::Function
+  rhs
+end
+
+@struct_hash_equal struct AndWhere <: AbstractCondition
+  conds::Vector{<:AbstractCondition}
+  AndWhere(conds::Vector{<:AbstractCondition}) = new(conds)
+  AndWhere(a, b) = new([a; b])
+end
+
+AndWhere(a::AndWhere, b) = AndWhere(a.conds, b)
+AndWhere(a, b::AndWhere) = AndWhere(a, b.conds)
+
+function Base.:&(a::S, b::T) where {T<:AbstractCondition, S<:AbstractCondition}
+  AndWhere(a, b)
+end
+
+@struct_hash_equal struct OrWhere <: AbstractCondition
+  conds::Vector{<:AbstractCondition}
+  OrWhere(conds::Vector{<:AbstractCondition}) = new(conds)
+  OrWhere(a, b) = new([a; b])
+end
+
+OrWhere(a::OrWhere, b) = OrWhere(a.conds, b)
+OrWhere(a, b::OrWhere) = OrWhere(a, b.conds)
+
+function Base.:|(a::S, b::T) where {T<:AbstractCondition, S<:AbstractCondition}
+  OrWhere(a, b)
+end
+
+""" """
+mutable struct ACSetSQLNode
+  const from::Symbol
+  cond::Union{Vector{<:AbstractCondition}, Nothing}
+  select
+  ACSetSQLNode(from::Symbol; cond=nothing, select=nothing) = new(from, cond, select)
+end
+
+function (w::WhereCondition)(node::ACSetSQLNode)
+  push!(node.cond, AndWhere([w]))
+  node
+end
+
+function (ac::AbstractCondition)(node::ACSetSQLNode)
+  push!(node.cond, ac)
+  node
+end
+
+function Base.:&(n::ACSetSQLNode, a::AbstractCondition)
+  n.cond = n.cond & a
+  n
+end
+
+function Base.:|(n::ACSetSQLNode, a::AbstractCondition)
+  n.cond = n.cond | a
```

Comment:
FYI just like there is `counter += 1`, you could be nifty here and use `n.cond |= a`, but it makes no real difference.